File: sandbox/demo/parent.py
import logging
import os
import signal

from sandbox.demo.child import child
from sandbox.demo.common import ACK_BYTE, READY_BYTE
from sandbox.demo.config import Config
from sandbox.demo.diag import log_identity
from sandbox.demo.idmap import write_idmaps


def parent(
    cfg: Config,
    *,
    logger: logging.Logger,
) -> int:
    host_uid = os.getuid()
    host_gid = os.getgid()

    logger.debug(f"host: uid={host_uid},gid={host_gid}")
    log_identity(logger, "parent (host namespace)", host_uid, host_gid)

    c2p_r, c2p_w = os.pipe()
    p2c_r, p2c_w = os.pipe()

    pid = os.fork()
    if pid == 0:
        # child
        os.close(c2p_r)
        os.close(p2c_w)
        rc = child(
            cfg,
            p2c_r,
            c2p_w,
            host_uid,
            host_gid,
        )
        os._exit(rc)

    # parent
    os.close(c2p_w)
    os.close(p2c_r)

    # wait for child readiness
    b = os.read(c2p_r, 1)
    if b != READY_BYTE:
        logger.error("child did not signal readiness; aborting")
        try:
            os.kill(pid, signal.SIGKILL)
        except Exception:
            pass
        return 1

    logger.debug(f"child ready; writing uid/gid maps for pid={pid}")
    try:
        write_idmaps(pid, host_uid, host_gid, logger=logger)
    except OSError as e:
        logger.error(f"failed to write uid/gid maps: {e}")
        try:
            os.kill(pid, signal.SIGKILL)
        except Exception:
            pass
        return 1

    # ACK child
    os.write(p2c_w, ACK_BYTE)

    # normal behavior: wait for child and exit
    _, status = os.waitpid(pid, 0)
    code = os.waitstatus_to_exitcode(status)

    logger.debug(f"child exit code: {code}")
    return code

File: sandbox/demo/parent/parent.go
//go:build linux

package parent

import (
	"log"
	"os"
	"syscall"

	"sandbox/demo/child"
	"sandbox/demo/common"
	"sandbox/demo/config"
	"sandbox/demo/diag"
	"sandbox/demo/idmap"
)

func Parent(cfg config.Config) int {
	hostUID := syscall.Getuid()
	hostGID := syscall.Getgid()

	log.Printf("[DEBUG] host: uid=%d,gid=%d", hostUID, hostGID)
	diag.LogIdentity("parent (host namespace)", hostUID, hostGID)

	c2pR, c2pW, _ := os.Pipe()
	p2cR, p2cW, _ := os.Pipe()

	pid, _, err := syscall.Syscall(syscall.SYS_FORK, 0, 0, 0)
	if err != 0 {
		log.Printf("[ERROR] fork failed: %v", err)
		return 1
	}

	if pid == 0 {
		c2pR.Close()
		p2cW.Close()
		rc := child.Child(cfg, p2cR.Fd(), c2pW.Fd(), hostUID, hostGID)
		os.Exit(rc)
	}

	c2pW.Close()
	p2cR.Close()

	buf := make([]byte, 1)
	if _, err := c2pR.Read(buf); err != nil || buf[0] != common.ReadyByte {
		log.Println("[ERROR] child did not signal readiness; aborting")
		_ = syscall.Kill(int(pid), syscall.SIGKILL)
		return 1
	}

	log.Printf("[DEBUG] child ready; writing uid/gid maps for pid=%d", pid)
	if err := idmap.WriteIDMaps(int(pid), hostUID, hostGID); err != nil {
		log.Printf("[ERROR] failed to write uid/gid maps: %v", err)
		_ = syscall.Kill(int(pid), syscall.SIGKILL)
		return 1
	}

	_, _ = p2cW.Write([]byte{common.AckByte})
	p2cW.Close()

	var status syscall.WaitStatus
	_, errWait := syscall.Waitpid(int(pid), &status, 0)
	if errWait != nil {
		return 1
	}

	log.Printf("[DEBUG] child exit code: %d", status.ExitStatus())
	return status.ExitStatus()
}
